clang/lib/Sema/SemaDeclAttr.cpp

```diff
@@ -2895,18 +2895,35 @@ static bool checkWorkGroupSizeValues(Sema &S, Decl *D, const ParsedAttr &Attr,
   if (const auto *A = D->getAttr<SYCLIntelMaxGlobalWorkDimAttr>())
     if (A->getNumber() == 0)
       Result &= checkZeroDim(A, WGSize[0], WGSize[1], WGSize[2],
-                             /*ReverseAttrs=*/ true);
+                             /*ReverseAttrs=*/true);
 
   if (const auto *A = D->getAttr<SYCLIntelMaxWorkGroupSizeAttr>()) {
-    if (!(WGSize[0] <= A->getXDim() && WGSize[1] <= A->getYDim() &&
+    if (S.getLangOpts().SYCLIsDevice &&
```

Comment:
There is a lot of code duplication around `S.getLangOpts().SYCLIsDevice`.
Consider the following improvement:
```
int XDim = WGSize[0];
int YDim = WGSize[1];
int ZDim = WGSize[2];
if (S.getLangOpts().SYCLIsDevice) {
  std::swap(XDim, ZDim); // add a good comment here
}
```
and then use `*Dim` variables instead of `WGSize`.